### PR TITLE
[5/n] Lora support cuda graph

### PR DESCRIPTION
**Pull request by yushengsu-thu**
## Motivation

MoE LoRA inference does not support CUDA graph because the forward path dynamically allocates intermediate tensors (torch.empty()) on every call. CUDA graph requires fixed tensor addresses between capture and replay, so these dynamic allocations break graph replay.

## Modifications

- Pre-allocate MoE CG buffers (Phase 1): Add init_cuda_graph_moe_buffers() to BaseLoRABackend, implemented in TritonLoRABackend and ChunkedSgmvLoRABackend. Called from ModelRunner before init_memory_pool() so memory profiling accounts for the buffers. All MoE LoRA layers share one buffer set since they execute sequentially.
- Reuse buffers in runner (Phase 2): TritonRunnerCoreWithLoRA.run() slices pre-allocated buffers instead of calling torch.empty() when in CG mode. Covers intermediate caches, alignment tensors, and output buffers.
- CG-aware LoRA path selection: Force the LoRA kernel path during capture (so kernels are recorded in the graph); skip LoRA path entirely when no adapter is active in the batch.
- In-place adapter_enabled update: _get_lora_info() reuses CG buffer with .zero_() + .index_fill_() instead of allocating a new tensor.
- Kernel dtype fix: Cast operand in fused_moe_lora_kernel to avoid dtype mismatch under mixed precision.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

File: python/sglang/srt/lora/layers.py [hunk 1]
```diff
@@ -838,42 +839,92 @@ def slice_lora_a_weights(self, A: torch.Tensor, tp_rank: int):
     def slice_lora_b_weights(self, B: torch.Tensor, tp_rank: int):
         return B
 
-    def slice_moe_lora_a_weights(
-        self, A: torch.Tensor, tp_rank: int, target_module: str
-    ) -> torch.Tensor:
+    def slice_moe_lora_a_weights(self, A, tp_rank: int, target_module: str):
```
[the remark this answers is not in the document]
Comment (yushengsu-thu) [reply python/sglang/srt/lora/layers.py]: done

File: python/sglang/srt/lora/layers.py [hunk 2]
```diff
@@ -838,42 +839,92 @@ def slice_lora_a_weights(self, A: torch.Tensor, tp_rank: int):
...
 
-    def slice_moe_lora_b_weights(
-        self, B: torch.Tensor, tp_rank: int, target_module: str
-    ) -> torch.Tensor:
+    @property
+    def _uses_interleaved_gate_up(self) -> bool:
+        cfg = getattr(self.base_layer, "moe_runner_config", None)
+        return getattr(cfg, "gemm1_alpha", None) is not None
+
+    def slice_moe_lora_b_weights(self, B, tp_rank: int, target_module: str):
```
[the remark this answers is not in the document]
Comment (yushengsu-thu) [reply python/sglang/srt/lora/layers.py]: done

File: python/sglang/srt/lora/mem_pool.py [hunk 3]
```diff
@@ -587,41 +613,77 @@ def load_lora_weight_tensor(
...
+                            load_lora_weight_tensor(buffer_view, w)
+                        elif isinstance(weights, dict) and len(weights) > 0:
+                            rep = next(iter(weights.values()))
+                            buffer_view = target_buffer[buffer_id, 0, :, :lora_rank]
+                            if rep is not None:
+                                rep = rep * lora_adapter.scaling
+                            load_lora_weight_tensor(buffer_view, rep)
+                        else:
+                            target_buffer[buffer_id].zero_()
+                    elif isinstance(weights, torch.Tensor) and weights.dim() == 3:
```
[the remark this answers is not in the document]
Comment (yushengsu-thu) [reply python/sglang/srt/lora/mem_pool.py:702]: done

File: test/registered/lora/test_lora_qwen3_8b_logprob_diff.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+        try:
+            cdata = torch.load(
+                os.path.join(adapter_path, "compare_sample_train_data.pt"),
+                weights_only=False,
+            )
+
+            base_logprobs = get_prompt_logprobs(engine, cdata["tokens"], lora_path=None)
+            logprobs = get_prompt_logprobs(engine, cdata["tokens"], lora_path="my_lora")
+
```
[the remark this answers is not in the document]
Comment (yushengsu-thu) [reply test/registered/lora/test_lora_qwen3_8b_logprob_diff.py:153]: done

File: python/sglang/srt/lora/layers.py [hunk 5]
```diff
@@ -750,26 +749,26 @@ def set_lora_info(
...
-            len(lora_ranks), dtype=torch.int32, device=lora_ranks.device
-        )
-        adapter_enabled.index_fill_(0, batch_info.weight_indices.long(), 1)
+        cg_buffers = getattr(self.lora_backend, "moe_cg_buffers", None)
+        if cg_buffers is not None and batch_info.use_cuda_graph:
+            adapter_enabled = cg_buffers["adapter_enabled"]
+            adapter_enabled.zero_()
+            adapter_enabled.index_fill_(
+                0, batch_info.weight_indices[: batch_info.bs].long(), 1
+            )
```
[the remark this answers is not in the document]
Comment (yushengsu-thu) [reply python/sglang/srt/lora/layers.py]: done

File: python/sglang/srt/lora/lora_moe_runners.py [hunk 6]
```diff
@@ -140,6 +148,28 @@ def run(
...
+            has_active_lora = True
+        else:
+            has_active_lora = (
+                (
+                    lora_info.adapter_enabled
+                    * (lora_info.lora_ranks > 0).to(lora_info.adapter_enabled.dtype)
+                )
+                .any()
+                .item()
+            )
```
[the remark this answers is not in the document]
Comment (yushengsu-thu) [reply python/sglang/srt/lora/lora_moe_runners.py]: done

File: python/sglang/srt/lora/utils.py [hunk 7]
```diff
@@ -220,6 +227,8 @@ def auto_detect_lora_target_modules(model: "torch.nn.Module") -> set:
             raw_names.add("down_proj")
         elif isinstance(module, ParallelLMHead):
             raw_names.add("lm_head")
+        elif isinstance(module, VocabParallelEmbedding):
```
Comment (Fridge003) [inline python/sglang/srt/lora/utils.py:230]: This operation is risky, since ParallelLMHead is a subclass of VocabParallelEmbedding. If a set of lora modules only include lm_head, embed_tokens will also be added mistakenly.

Can be checked later

File: python/sglang/srt/lora/lora_moe_runners.py
Comment (Fridge003) [inline python/sglang/srt/lora/lora_moe_runners.py:72]: We had better rename `LoRAInfo` to `MoELoRAInfo`. For future PR

**Review by Fridge003 — APPROVED** [review]
